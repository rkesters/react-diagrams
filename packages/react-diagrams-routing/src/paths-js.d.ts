declare module 'paths-js/path' {
	export interface Instruction {
		params: number[];
		command: 'default' | 'V' | 'H' | 'A' | 'C' | 'Z';
	}

	export interface Point {
		x: number;
		y: number;
	}

	export interface CurveTo {
		x1: number;
		y1: number;
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface SmoothCurveTo {
		x2: number;
		y2: number;
		x: number;
		y: number;
	}

	export interface QCurveTo {
		x1: number;
		y1: number;
		x: number;
		y: number;
	}

	export interface Arc {
		rx: number;
		ry: number;
		xrot: number;
		largeArcFlag: number;
		sweepFlag: number;
		x: number;
		y: number;
	}

	export interface Translate {
		dx: number;
		dy: number;
	}

	export interface Rotate {
		angle: number;
		rx: number;
		ry: number;
	}

	export interface Scale {
		sx: number;
		sy: number;
	}

	export interface Shear {

	}

	export interface IPath {
		moveto(point: Point): IPath;
		moveto(x: number, y: number): IPath;
		moveto(x: number | Point, y?: number): IPath;

		lineto(point: Point): IPath;
		lineto(x: number, y: number): IPath;
		lineto(x: number | Point, y?: number): IPath;

		hlineto(point: Point): IPath;
		hlineto(x: number, y: number): IPath;
		hlineto(x: number | Point, y?: number): IPath;

		vlineto(point: Point): IPath;
		vlineto(x: number, y: number): IPath;
		vlineto(x: number | Point, y?: number): IPath;

		closepath(): IPath;

		curveto(curve: CurveTo): IPath;
		curveto(x1: number, y1: number, x2: number, y2: number, x: number): IPath;
		curveto(x1: number | CurveTo, y1?: number, x2?: number, y2?: number, x?: number): IPath;

		smoothcurveto(point: SmoothCurveTo): IPath;
		smoothcurveto(x2: number, y2: number, x: number): IPath;
		smoothcurveto(x2: number | SmoothCurveTo, y2?: number, x?: number): IPath;

		qcurveto(point: QCurveTo): IPath;
		qcurveto(x1: number, y1: number, x: number): IPath;
		qcurveto(x1: number | QCurveTo, y1?: number, x?: number): IPath;

		smoothqcurveto(point: Point): IPath;
		smoothqcurveto(x: number, y: number): IPath;
		smoothqcurveto(x: number | Point, y?: number): IPath;

		arc(point: Arc): IPath;
		arc(rx: number, ry: number, xrot: number, largeArcFlag: number, sweepFlag: number, x: number, y: number): IPath;
		arc(
			rx: number | Arc,
			ry?: number,
			xrot?: number,
			largeArcFlag?: number,
			sweepFlag?: number,
			x?: number,
			y?: number
		): IPath;

		translate(point: Translate): IPath;
		translate(dx: number, dy: number): IPath;
		translate(dx: number | Translate, dy?: number): IPath;

		rotate(point: Rotate): IPath;
		rotate(angle: number, rx: number, ry: number): IPath;
		rotate(angle: number | Rotate, rx: number, ry: number): IPath;

		scale(scaleBy: Scale): IPath;
		scale(sx: number, sy: number): IPath;
		scale(sx: number | Scale, sy?: number): IPath;

		shearX(angle: Scale): IPath;
		shearX(angle: number): IPath;
		shearX(angle: Scale | number): IPath;

		shearY(angle: Scale): IPath;
		shearY(angle: number): IPath;
		shearY(angle: Scale | number): IPath;

		print(): string;
		toString(): string;
		points() : [number, number][];
		instructions() : Instruction[];

		connect(path: IPath): IPath;
	}

	export default function Path(init?: Instruction[]): IPath;
}
