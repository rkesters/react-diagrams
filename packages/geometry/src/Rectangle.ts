import { isNumber } from 'lodash';
import { Point } from './Point';
import { Polygon } from './Polygon';

export class Rectangle extends Polygon {
	constructor(tl: Point, tr: Point, br: Point, bl: Point);
	constructor(position: Point, width: number, height: number);
	constructor(x?: number, y?: number, width?: number, height?: number);

	constructor(a: Point | number = 0, b: Point | number  = 0, c: Point | number  = 0, d: Point | number  = 0) {
		if (Point.isa(a) && Point.isa(b) && Point.isa(c) && Point.isa(d)) {
			super([a, b, c, d]);
			return;
		} else if (Point.isa(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super([a, new Point(a.x + b, a.y), new Point(a.x + b, a.y + c), new Point(a.x, a.y + c)]);
			return;
		} else if (isNumber(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super(Rectangle.pointsFromBounds(a, b, c, d));
			return;
		}

		throw new Error(`Rectangle call with invaild arguements: ${arguments.length}`);
	}

	static pointsFromBounds(x: number, y: number, width: number, height: number): Point[] {
		return [new Point(x, y), new Point(x + width, y), new Point(x + width, y + height), new Point(x, y + height)];
	}

	updateDimensions(x: number, y: number, width: number, height: number) {
		this.points = Rectangle.pointsFromBounds(x, y, width, height);
	}

	setPoints(points: Point[]) {
		if (points.length !== 4) {
			throw 'Rectangles must always have 4 points';
		}
		super.setPoints(points);
	}

	containsPoint(point: Point) {
		const tl = this.getTopLeft();
		const br = this.getBottomRight();

		return point.x >= tl.x && point.x <= br.x && point.y >= tl.y && point.y <= br.y;
	}

	getWidth(): number {
		return Math.sqrt(
			Math.pow(this.getTopLeft().x - this.getTopRight().x, 2) + Math.pow(this.getTopLeft().y - this.getTopRight().y, 2)
		);
	}

	getHeight(): number {
		return Math.sqrt(
			Math.pow(this.getBottomLeft().x - this.getTopLeft().x, 2) +
				Math.pow(this.getBottomLeft().y - this.getTopLeft().y, 2)
		);
	}

	getTopMiddle(): Point {
		return Point.middlePoint(this.getTopLeft(), this.getTopRight());
	}

	getBottomMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getBottomRight());
	}

	getLeftMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getTopLeft());
	}

	getRightMiddle(): Point {
		return Point.middlePoint(this.getBottomRight(), this.getTopRight());
	}

	getTopLeft(): Point {
		return this.points[0];
	}

	getTopRight(): Point {
		return this.points[1];
	}

	getBottomRight(): Point {
		return this.points[2];
	}

	getBottomLeft(): Point {
		return this.points[3];
	}
}
