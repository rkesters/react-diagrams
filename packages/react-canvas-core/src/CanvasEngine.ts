import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
