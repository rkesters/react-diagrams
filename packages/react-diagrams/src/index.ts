import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
