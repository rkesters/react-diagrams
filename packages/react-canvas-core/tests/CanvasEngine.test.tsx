import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import React from 'react';
import { CanvasEngine } from '../src/CanvasEngine';
import { CanvasModel } from '../src/entities/canvas/CanvasModel';
import { TestLayer, TestReactFactory } from './support/impls/TestLayer';

describe('CanvasEngine', () => {
	test.each([
		{
			name: 'Provide Options',
			options: {
				registerDefaultDeleteItemsAction: false,
				registerDefaultZoomCanvasAction: false,
				repaintDebounceMs: 10
			}
		},
		{
			name: 'Provide Default Options',
			options: {}
		},
		{
			name: 'Provide Undefined Options',
			options: undefined
		}
	])('constructor: $name', (tc) => {
		const engine = new CanvasEngine(tc.options);

		//TODO: typing does not all for it to be null
		expect(engine.getModel()).toBeNull();

		expect(engine.getActionEventBus()).toBeDefined();
		expect(engine.getStateMachine()).toBeDefined();
		expect(engine.getLayerFactories()).toBeDefined();
		expect(engine.getOptions()).toMatchSnapshot();
	});

	test('getter/setters', () => {
		const engine = new CanvasEngine();
		const { container, debug } = render(
			<div
				id="canvas"
				style={{
					padding: '20px',
					margin: '50px auto',
					whiteSpace: 'nowrap',
					width: '10px',
					height: '20px',
					overflow: 'hidden'
				}}>
				<div id="c" style={{ width: '100px', height: '200px', overflow: 'hidden' }}>
					Love is all you need
				</div>
			</div>
		);
		const canvas: HTMLDivElement = container.querySelector('#canvas');
		engine.setCanvas(canvas);
		expect(engine.getCanvas()).toEqual(canvas);

		engine.setModel(new CanvasModel());
		expect(engine.getModel()).toBeDefined();

		engine.setModel(null);
		expect(engine.getModel()).toBeNull();

		engine.setModel(new CanvasModel());

		engine.getModel().setOffset(10, 20);

		expect(engine.getRelativeMousePoint({ clientX: 50, clientY: 60 })).toMatchInlineSnapshot(`
		Point {
		  "x": 40,
		  "y": 40,
		}
	`);

		expect(() => engine.getFactoryForLayer('badlayer')).toThrowError();
		engine.getLayerFactories().registerFactory(new TestReactFactory('test'));
		expect(engine.getFactoryForLayer('test')).toBeDefined();
		expect(engine.getFactoryForLayer(new TestLayer())).toBeDefined();
		expect(engine.getMouseElement({} as any)).toBeNull();

		Object.defineProperty(HTMLElement.prototype, 'scrollWidth', { configurable: true, value: 500 });
		Object.defineProperty(HTMLElement.prototype, 'scrollHeight', { configurable: true, value: 500 });
		Object.defineProperty(HTMLElement.prototype, 'clientWidth', { configurable: true, value: 200 });
		Object.defineProperty(HTMLElement.prototype, 'clientHeight', { configurable: true, value: 200 });
		engine.zoomToFit();
		expect(engine.getModel().getOffsetX()).toEqual(0);
		expect(engine.getModel().getOffsetY()).toEqual(0);
		expect(engine.getModel().getZoomLevel()).toEqual(40);


		expect(engine.repaintCanvas(true)).toHaveProperty('then');
		expect(engine.repaintCanvas()).toBeUndefined();
	});
});
