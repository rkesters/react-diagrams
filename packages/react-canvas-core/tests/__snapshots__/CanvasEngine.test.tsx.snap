// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasEngine constructor: Provide Default Options 1`] = `
Object {
  "registerDefaultDeleteItemsAction": true,
  "registerDefaultZoomCanvasAction": true,
  "repaintDebounceMs": 0,
}
`;

exports[`CanvasEngine constructor: Provide Options 1`] = `
Object {
  "registerDefaultDeleteItemsAction": false,
  "registerDefaultZoomCanvasAction": false,
  "repaintDebounceMs": 10,
}
`;

exports[`CanvasEngine constructor: Provide Undefined Options 1`] = `
Object {
  "registerDefaultDeleteItemsAction": true,
  "registerDefaultZoomCanvasAction": true,
  "repaintDebounceMs": 0,
}
`;
